File: erpnext_enhancements/chat/gchat/markdown.py
# Copyright (c) 2026, Sapphire Fountains and contributors
# For license information, please see license.txt

"""Markdown → Google Chat's own text formatting. The third of the three renderers.

**The bug this exists to fix, and why it looked intermittent.** Google Chat's bold is a
**single** asterisk (``*bold*``), not a double one. Triton emits CommonMark, so
``**Incompressibility:**`` reached the Chat client as ``*`` + ``*Incompressibility:*`` + ``*``
— which Chat renders as the phrase **in bold with a stray asterisk glued to it**. To a reader
that looks like the formatting works sometimes and produces a spare character other times,
which is exactly how it was reported. It is neither: it is one deterministic mistranslation.

Chat's whole formatting vocabulary is small and this module targets it exactly:

===================  ==========================  ==========================================
CommonMark           Google Chat                 Note
===================  ==========================  ==========================================
``**b**`` / ``__b__``  ``*b*``                   the reported bug
``*i*`` / ``_i_``      ``_i_``                   asterisk-italic must become underscore
``~~s~~``              ``~s~``
`````c`````            unchanged                 already Chat's syntax
fenced block           unchanged                 already Chat's syntax
``# Heading``          ``*Heading*``             Chat has no headings; bold is the closest
``- item``             ``• item``                Chat has no list syntax; the bullet is text
``1. item``            unchanged                 already reads as a list
``[label](url)``       ``label: url``            Chat autolinks bare URLs; ``<u|l>`` is Slack
``> quoted``           ``> quoted``              no blockquote in Chat; the prefix is the cue
a table                a monospace block         no table syntax; the block keeps the columns
===================  ==========================  ==========================================

WHY THIS IS A TOKENISER AND NOT A CHAIN OF ``re.sub`` CALLS
===========================================================

The obvious implementation is three substitutions, and it is wrong in a way that only shows up
on real input. Turning ``**b**`` into ``*b*`` and *then* turning ``*i*`` into ``_i_`` converts
the output of the first rule with the second, so every bold word silently becomes italic. Any
ordering has a case like it, because the source and target alphabets overlap.

So this walks the string once, longest-delimiter-first, and never re-examines what it has
already emitted — the same shape as the browser-side renderer in
``public/js/chat/markdown.js``, for the same reason.

**Code spans are extracted first and restored last.** Their contents are literal by definition:
``` `**not bold**` ``` must survive with both asterisks, and a substitution pass would eat them.

Indentation is tabs, per ``CLAUDE.md``.
"""

from __future__ import annotations

import re
from typing import Final

#: Fenced blocks and inline code, matched together so the longer form wins. Extracted before
#: any conversion and put back afterwards, untouched.
_CODE_RE: Final = re.compile(r"(```.*?```|`[^`\n]+`)", re.S)

#: Inline rules, tried longest-delimiter-first at each position. Order is load-bearing: ``**``
#: before ``*``, or bold degrades to italic-plus-a-stray-asterisk — the same class of mistake
#: this module exists to undo.
_INLINE_RULES: Final = (
	# Non-greedy, and NOT `[^*]*`: the inner run must be allowed to contain single asterisks
	# so that `**bold with *inner* italic**` matches the outer pair. Non-greedy is what stops
	# `**a** and **b**` being read as one span from the first `**` to the last.
	(re.compile(r"\*\*(\S(?:.*?\S)?)\*\*"), "*{}*"),
	(re.compile(r"__(\S(?:.*?\S)?)__"), "*{}*"),
	(re.compile(r"~~(\S(?:.*?\S)?)~~"), "~{}~"),
	(re.compile(r"\*(\S(?:[^*]*\S)?)\*"), "_{}_"),
)

#: ``[label](url)``. Chat has no link syntax for plain-text messages — ``<url|label>`` is
#: Slack's, and it arrives in Chat as those literal characters — so the label and the URL are
#: both emitted and Chat's own autolinker makes the URL clickable.
_LINK_RE: Final = re.compile(r"\[([^\]\n]*)\]\(([^)\s]+)\)")

_HEADING_RE: Final = re.compile(r"^(#{1,6})\s+(.*)$")
_QUOTE_RE: Final = re.compile(r"^\s*>\s?(.*)$")
_TABLE_ROW_RE: Final = re.compile(r"^\s*\|(.+)\|\s*$")

#: The alignment row is what distinguishes a table from two lines that happen to contain
#: pipes. Markdown requires it; requiring it here means a message about ``a | b`` is never
#: silently reinterpreted as a one-column table.
_TABLE_DIVIDER_RE: Final = re.compile(r"^\s*\|(?:\s*:?-{1,}:?\s*\|)+\s*$")
_BULLET_RE: Final = re.compile(r"^(\s*)[-*+]\s+(.*)$")

#: What a bullet becomes. A real bullet character rather than ``-``, because the point is that
#: it reads as a list in a client that has no list markup.
_BULLET: Final = "• "

#: Underscore emphasis does not open inside a word: ``snake_case_name`` is one identifier, and
#: Triton emits identifiers constantly. Matches the browser renderer's rule and CommonMark's.
_WORD_CHAR: Final = re.compile(r"[A-Za-z0-9]")


def to_google_chat(source: str | None) -> str:
	"""Rewrite CommonMark into Google Chat's formatting. Pure; no I/O, no settings.

	Anything the table above does not name passes through as the characters the model typed,
	which is the correct failure mode: an unconverted construct is a cosmetic miss, while a
	half-converted one is the stray-asterisk bug in a new costume.
	"""
	text = str(source or "")
	if not text:
		return ""

	spans: list[str] = []

	def _stash(match: re.Match[str]) -> str:
		spans.append(match.group(0))
		return f"\x00{len(spans) - 1}\x00"

	text = _CODE_RE.sub(_stash, text)
	text = _convert_tables(text)
	text = "\n".join(_convert_line(line) for line in text.split("\n"))

	for index, span in enumerate(spans):
		text = text.replace(f"\x00{index}\x00", span)
	return text


def _convert_tables(text: str) -> str:
	"""Markdown tables → a monospace block with padded columns.

	**Google Chat has no table syntax**, so the choice is between losing the structure and
	faking it. It has a monospace block, which preserves the one thing a table is *for*:
	columns that line up. Padding the cells and wrapping the result in a fence is the closest
	Chat can get, and it reads correctly on a phone where a wide table would not anyway.

	The header separator becomes a row of dashes rather than being dropped, because without it
	the header stops looking like a header.

	A table is recognised only when the alignment row is present and immediately follows the
	header — markdown requires it, and requiring it here means a message about ``a | b`` is
	never silently reinterpreted as a one-column table.
	"""
	lines = text.split("\n")
	out: list[str] = []
	index = 0

	while index < len(lines):
		is_table = (
			_TABLE_ROW_RE.match(lines[index])
			and index + 1 < len(lines)
			and _TABLE_DIVIDER_RE.match(lines[index + 1])
		)
		if not is_table:
			out.append(lines[index])
			index += 1
			continue

		rows = [_table_cells(lines[index])]
		index += 2  # the header and its alignment row
		while index < len(lines) and _TABLE_ROW_RE.match(lines[index]) and not _TABLE_DIVIDER_RE.match(lines[index]):
			rows.append(_table_cells(lines[index]))
			index += 1

		width = max(len(row) for row in rows)
		rows = [row + [""] * (width - len(row)) for row in rows]
		# The emphasis inside a cell is stripped rather than converted: a monospace block in
		# Chat renders its contents literally, so `*bold*` inside one is asterisks on screen.
		rows = [[_strip_emphasis(cell) for cell in row] for row in rows]
		sizes = [max(len(row[col]) for row in rows) for col in range(width)]

		rendered = ["```"]
		for position, row in enumerate(rows):
			rendered.append("  ".join(cell.ljust(sizes[col]) for col, cell in enumerate(row)).rstrip())
			if position == 0:
				rendered.append("  ".join("-" * sizes[col] for col in range(width)))
		rendered.append("```")
		out.extend(rendered)

	return "\n".join(out)


def _table_cells(line: str) -> list[str]:
	return [cell.strip() for cell in line.strip().strip("|").split("|")]


def _strip_emphasis(cell: str) -> str:
	"""Remove emphasis delimiters from a cell bound for a monospace block.

	Converting them would be worse than dropping them: Chat renders a monospace block
	literally, so a converted ``*bold*`` arrives as two visible asterisks inside a table that
	is trying to line its columns up.
	"""
	for pattern in (r"\*\*(.+?)\*\*", r"__(.+?)__", r"~~(.+?)~~", r"\*(.+?)\*", r"`(.+?)`"):
		cell = re.sub(pattern, r"\1", cell)
	return cell


def _convert_line(line: str) -> str:
	quote = _QUOTE_RE.match(line)
	if quote:
		# Chat renders no blockquote. The `>` prefix is kept as literal text because it is the
		# convention every reader already knows, and dropping it would silently merge quoted
		# text into the surrounding message — which is the one thing a quote must not do.
		return f"> {_convert_inline(quote.group(1))}" if quote.group(1).strip() else ">"

	heading = _HEADING_RE.match(line)
	if heading:
		# Chat has no headings. Bold is the closest thing it has, and a heading rendered as
		# body text loses the only cue that a section started.
		return f"*{_convert_inline(heading.group(2))}*" if heading.group(2).strip() else ""

	bullet = _BULLET_RE.match(line)
	if bullet:
		return f"{bullet.group(1)}{_BULLET}{_convert_inline(bullet.group(2))}"

	return _convert_inline(line)


def _convert_inline(text: str) -> str:
	text = _LINK_RE.sub(lambda m: f"{m.group(1)}: {m.group(2)}" if m.group(1) else m.group(2), text)

	out: list[str] = []
	index = 0
	while index < len(text):
		match = _first_match(text, index)
		if match is None:
			out.append(text[index:])
			break
		rule_match, template = match
		out.append(text[index : rule_match.start()])
		out.append(template.format(_convert_inline(rule_match.group(1))))
		index = rule_match.end()
	return "".join(out)


def _first_match(text: str, start: int) -> tuple[re.Match[str], str] | None:
	"""The earliest rule match at or after ``start``, ties going to the earlier rule.

	Ties matter: at the same offset ``**`` and ``*`` both match, and the tuple order is what
	makes bold win. ``_italic_`` is handled by the *underscore* branch below rather than by a
	rule, because it needs the intraword guard and because its output is its input — Chat and
	CommonMark spell italic the same way, so the only work is deciding whether it is one.
	"""
	best: tuple[re.Match[str], str] | None = None
	for pattern, template in _INLINE_RULES:
		found = pattern.search(text, start)
		if found and (best is None or found.start() < best[0].start()):
			best = (found, template)

	underscore = _underscore_italic(text, start)
	if underscore and (best is None or underscore.start() < best[0].start()):
		best = (underscore, "_{}_")
	return best


def _underscore_italic(text: str, start: int) -> re.Match[str] | None:
	"""``_italic_``, but only where it is not inside a word.

	Scans forward past rejected hits rather than giving up on the first one: in
	``a_b and _real_ italic`` the intraword run must be skipped and the genuine emphasis still
	found.
	"""
	pattern = re.compile(r"_(\S(?:[^_]*\S)?)_")
	position = start
	while position < len(text):
		found = pattern.search(text, position)
		if not found:
			return None
		before = text[found.start() - 1] if found.start() > 0 else ""
		after = text[found.end()] if found.end() < len(text) else ""
		if not _WORD_CHAR.match(before or "") and not _WORD_CHAR.match(after or ""):
			return found
		position = found.start() + 1
	return None

File: erpnext_enhancements/public/js/chat/markdown.js
/**
 * The one markdown renderer, shared by the chat SPA and the floating widget.
 *
 * WHY THIS EXISTS RATHER THAN `frappe.markdown`
 * =============================================
 *
 * Three surfaces rendered Triton's answers three different ways, and all three were wrong:
 *
 * 1. The **SPA** built message bodies entirely with `textContent`, so an answer arrived as
 *    literal `**bold**` and `- item`. No markdown renderer existed on that page at all.
 * 2. The **widget** called `frappe.markdown`, which in v16 is `new showdown.Converter()` with
 *    **no sanitiser** — and which lives in the Desk bundle. On a Desk page it rendered; on
 *    `/chat`, where `frappe`/`showdown`/`jQuery` are absent, it threw and fell back **silently**
 *    to escaped raw text. Same widget, two different outputs, no error either way.
 * 3. **Google Chat** got the raw markdown. Chat's own bold is a *single* asterisk, so
 *    `**Incompressibility:**` arrived bold with a stray `*` glued to it. That transform is
 *    server-side (`chat/gchat/markdown.py`); this module is the browser half.
 *
 * So the fix could not be "call `frappe.markdown` in more places": that spreads an unsanitised
 * `innerHTML` sink to two more surfaces and still does nothing on `/chat`.
 *
 * THE RULE THIS FILE KEEPS
 * ========================
 *
 * **Nodes, never HTML strings.** Every element here is `createElement` and every piece of text
 * is `textContent`. There is no `innerHTML` in this module and there must never be one — the
 * input is model output, which is untrusted twice over: it can carry prompt-injected content,
 * and it quotes employees' own messages back verbatim. `citations.js` states the same rule as
 * its rule 3, and this renderer is downstream of the same threat.
 *
 * A consequence worth stating: **this is a renderer, not a parser.** It handles the constructs
 * Triton actually emits — bold, italic, inline code, fenced code, bullet and numbered lists,
 * links, headings — and anything else falls through as literal text. That is the correct
 * failure mode. An unhandled construct rendering as the characters the model typed is a
 * cosmetic miss; an unhandled construct rendering as markup is a vulnerability.
 *
 * Tables and blockquotes are deliberately **not** supported (settled 2026-08-13). Add them by
 * extending `renderBlocks`, with corpus rows, never by reaching for a general parser.
 */

import { isSafeUrl } from "./citations.js";

/** Fenced code: ```lang\n...\n``` — `lang` captured so it can be shown, never executed. */
const FENCE_RE = /^```([A-Za-z0-9_+-]*)\s*$/;

/** `# ` through `###### `. Deeper hashes are not headings in any dialect worth matching. */
const HEADING_RE = /^(#{1,6})\s+(.*)$/;

/** `- `, `* ` or `+ ` with optional leading indent. The indent is captured for nesting depth. */
const BULLET_RE = /^(\s*)[-*+]\s+(.*)$/;

/** `1. ` / `1) `, same shape. */
const ORDERED_RE = /^(\s*)(\d{1,9})[.)]\s+(.*)$/;

/**
 * Inline constructs, in the order they are tried. **Order is load-bearing**: `**` must be
 * attempted before `*`, or bold renders as an italic wrapping a literal asterisk — which is
 * precisely the bug Google Chat exhibits for the opposite reason.
 *
 * Code is first because its contents are literal: `` `**not bold**` `` must stay asterisks.
 */
const INLINE_RULES = [
	{ kind: "code", re: /`([^`\n]+)`/ },
	{ kind: "link", re: /\[([^\]\n]*)\]\(([^)\s]+)\)/ },
	// Non-greedy, and NOT `[^*]*`: the inner run must allow single asterisks so that
	// `**bold with *inner* italic**` matches the outer pair. Non-greedy stops
	// `**a** and **b**` being read as one span from the first `**` to the last.
	{ kind: "bold", re: /\*\*(\S(?:.*?\S)?)\*\*/ },
	{ kind: "bold", re: /__(\S(?:.*?\S)?)__/, wordBoundary: true },
	{ kind: "italic", re: /\*(\S(?:[^*]*\S)?)\*/ },
	{ kind: "italic", re: /_(\S(?:[^_]*\S)?)_/, wordBoundary: true },
	{ kind: "strike", re: /~~(\S(?:.*?\S)?)~~/ },
];

/**
 * Word characters either side of which an underscore does **not** open emphasis.
 *
 * `snake_case_name` is one identifier, not a word with an italic in the middle — and Triton
 * emits identifiers constantly: field names, DocType scrub names, Python functions. CommonMark
 * draws exactly this distinction, allowing intraword `*` and forbidding intraword `_`, and the
 * reason is this case.
 *
 * Checked with an explicit character test rather than a lookbehind in the pattern: lookbehind
 * only reached Safari in 16.4, and the failure mode of an unsupported group is the regex
 * throwing at parse time — which would take the whole renderer down on that browser, silently,
 * for everybody on it.
 */
const WORD_CHAR = /[A-Za-z0-9]/;

/** Bare URLs, so a pasted link is clickable without markdown syntax around it. */
const BARE_URL_RE = /\bhttps?:\/\/[^\s<>()[\]]+/;

/** The tag each inline kind becomes. `code` is `<code>`; the rest are semantic emphasis. */
const INLINE_TAG = { bold: "strong", italic: "em", strike: "s", code: "code" };

/**
 * Render markdown into `container`, as nodes.
 *
 * @param {Element} container - emptied, then filled. Never assigned `innerHTML`.
 * @param {string} source - the markdown.
 * @param {object} [opts]
 * @param {Document} [opts.doc] - injected for tests, which run under a fake document in plain
 *   `node` with no DOM. Defaults to the real one.
 * @returns {Element} the container, for chaining.
 */
export function renderMarkdown(container, source, opts) {
	const doc = (opts && opts.doc) || (typeof document !== "undefined" ? document : null);
	if (!container || !doc) return container;

	while (container.firstChild) container.removeChild(container.firstChild);
	for (const node of renderBlocks(doc, String(source == null ? "" : source))) {
		container.appendChild(node);
	}
	return container;
}

/**
 * The block pass: split into paragraphs, lists, fenced code and headings.
 *
 * Exported for the test suite, which asserts the block structure directly rather than through
 * a container — a renderer whose only assertion is "the container has children" is a renderer
 * with no test.
 */
export function renderBlocks(doc, source) {
	const lines = String(source == null ? "" : source).replace(/\r\n?/g, "\n").split("\n");
	const out = [];
	let i = 0;

	while (i < lines.length) {
		const line = lines[i];

		const fence = line.match(FENCE_RE);
		if (fence) {
			// Everything up to the closing fence is literal, including markdown syntax. An
			// unterminated fence runs to the end of the message rather than swallowing the
			// document into a parse error — the model does truncate mid-block.
			const body = [];
			i += 1;
			while (i < lines.length && !FENCE_RE.test(lines[i])) {
				body.push(lines[i]);
				i += 1;
			}
			i += 1; // the closing fence, or the end
			const pre = doc.createElement("pre");
			pre.className = "ee-md-code";
			const code = doc.createElement("code");
			if (fence[1]) code.className = `language-${fence[1]}`;
			code.textContent = body.join("\n");
			pre.appendChild(code);
			out.push(pre);
			continue;
		}

		const heading = line.match(HEADING_RE);
		if (heading) {
			// Capped at h4 in the DOM regardless of hash count: these render inside a chat
			// bubble, and an <h1> in a bubble is a layout bug wearing semantics. The level is
			// kept as a class so styling can still distinguish them.
			const level = Math.min(heading[1].length, 6);
			const el = doc.createElement(`h${Math.min(level + 2, 6)}`);
			el.className = `ee-md-h ee-md-h${level}`;
			appendInline(doc, el, heading[2]);
			out.push(el);
			i += 1;
			continue;
		}

		if (BULLET_RE.test(line) || ORDERED_RE.test(line)) {
			const [list, consumed] = renderList(doc, lines, i);
			out.push(list);
			i = consumed;
			continue;
		}

		if (!line.trim()) {
			i += 1;
			continue;
		}

		// A paragraph runs to the next blank line or the next block construct. Single newlines
		// inside it become <br>, because a chat answer's line breaks are meaningful — markdown's
		// "join the lines" rule reads as a bug when the model formatted a short list by hand.
		const para = doc.createElement("p");
		para.className = "ee-md-p";
		let first = true;
		while (
			i < lines.length &&
			lines[i].trim() &&
			!FENCE_RE.test(lines[i]) &&
			!HEADING_RE.test(lines[i]) &&
			!BULLET_RE.test(lines[i]) &&
			!ORDERED_RE.test(lines[i])
		) {
			if (!first) para.appendChild(doc.createElement("br"));
			appendInline(doc, para, lines[i]);
			first = false;
			i += 1;
		}
		out.push(para);
	}

	return out;
}

/**
 * One list, from `start`, returning `[element, indexAfter]`.
 *
 * Nesting is by leading whitespace, and it is deliberately shallow: a deeper item becomes a
 * nested list, and anything past two levels flattens into the second. Chat bubbles are narrow
 * and the model's own indentation is not reliable enough to trust a fourth level of it.
 */
function renderList(doc, lines, start) {
	const ordered = ORDERED_RE.test(lines[start]);
	const root = doc.createElement(ordered ? "ol" : "ul");
	root.className = "ee-md-list";

	let i = start;
	let current = root;
	let currentDepth = 0;

	while (i < lines.length) {
		const bullet = lines[i].match(BULLET_RE);
		const numbered = lines[i].match(ORDERED_RE);
		if (!bullet && !numbered) break;

		const indent = (bullet ? bullet[1] : numbered[1]).replace(/\t/g, "  ").length;
		const text = bullet ? bullet[2] : numbered[3];
		const depth = indent >= 2 ? 1 : 0;

		if (depth > currentDepth) {
			const nested = doc.createElement(numbered ? "ol" : "ul");
			nested.className = "ee-md-list is-nested";
			(current.lastChild || current).appendChild(nested);
			current = nested;
			currentDepth = 1;
		} else if (depth < currentDepth) {
			current = root;
			currentDepth = 0;
		}

		const li = doc.createElement("li");
		li.className = "ee-md-li";
		appendInline(doc, li, text);
		current.appendChild(li);
		i += 1;
	}

	return [root, i];
}

/**
 * The inline pass. Appends text and emphasis nodes for one line into `parent`.
 *
 * Exported because it is the half most worth testing directly: nesting, unmatched delimiters,
 * and the ordering between `**` and `*` all live here.
 */
export function appendInline(doc, parent, text) {
	const source = String(text == null ? "" : text);
	if (!source) return parent;

	// The earliest match across all rules wins, so `a *b* and `c`` splits at the emphasis
	// rather than at whichever rule happens to be listed first. Ties go to the earlier rule,
	// which is what puts `**` ahead of `*` at the same offset.
	let best = null;
	for (const rule of INLINE_RULES) {
		const match = matchRule(rule, source);
		if (match && (best === null || match.index < best.match.index)) {
			best = { rule, match };
		}
	}
	const url = BARE_URL_RE.exec(source);
	if (url && (best === null || url.index < best.match.index)) {
		best = { rule: { kind: "bare" }, match: url };
	}

	if (!best) {
		parent.appendChild(doc.createTextNode(source));
		return parent;
	}

	const { rule, match } = best;
	if (match.index > 0) appendInline(doc, parent, source.slice(0, match.index));

	if (rule.kind === "bare") {
		parent.appendChild(anchor(doc, match[0], match[0]));
	} else if (rule.kind === "link") {
		parent.appendChild(anchor(doc, match[1] || match[2], match[2]));
	} else if (rule.kind === "code") {
		// Literal by definition: no recursion into the contents, which is what makes
		// `` `**not bold**` `` render as the characters the model typed.
		const el = doc.createElement("code");
		el.className = "ee-md-inline-code";
		el.textContent = match[1];
		parent.appendChild(el);
	} else {
		const el = doc.createElement(INLINE_TAG[rule.kind]);
		appendInline(doc, el, match[1]);
		parent.appendChild(el);
	}

	const rest = source.slice(match.index + match[0].length);
	if (rest) appendInline(doc, parent, rest);
	return parent;
}

/**
 * The first match of `rule` in `source` that clears the rule's own boundary condition.
 *
 * Scans forward rather than giving up on the first rejected hit: in `a_b and _real_ italic`,
 * the intraword `_b and _` must be skipped and the genuine `_real_` still found.
 */
function matchRule(rule, source) {
	if (!rule.wordBoundary) return rule.re.exec(source);

	let offset = 0;
	while (offset < source.length) {
		const match = rule.re.exec(source.slice(offset));
		if (!match) return null;

		const start = offset + match.index;
		const end = start + match[0].length;
		const before = start > 0 ? source[start - 1] : "";
		const after = end < source.length ? source[end] : "";
		if (!WORD_CHAR.test(before) && !WORD_CHAR.test(after)) {
			match.index = start;
			return match;
		}
		offset = start + 1;
	}
	return null;
}

/**
 * An anchor, or inert text when the URL is not safe.
 *
 * Routed through `isSafeUrl` — the parser-based check, not a prefix test — so `javascript:`,
 * `data:`, a credentials trick and the `/\host` origin escape all render as visible text rather
 * than as something clickable. A model can be talked into emitting any of them.
 */
function anchor(doc, label, href) {
	const text = String(label == null ? "" : label);
	if (!isSafeUrl(href)) {
		const span = doc.createElement("span");
		span.className = "ee-md-link is-unsafe";
		span.textContent = text;
		return span;
	}
	const a = doc.createElement("a");
	a.className = "ee-md-link";
	a.textContent = text;
	a.setAttribute("href", href);
	if (/^https?:\/\//i.test(String(href))) {
		a.setAttribute("target", "_blank");
		a.setAttribute("rel", "noopener noreferrer");
	}
	return a;
}
